# reverse-zMAP

reverse-zMAP module is primarily designed for handling MS runs with relatively large sample sizes. 
 In practice,large-scale proteomic studies have frequently applied the strategy of adding a biologically identical
 reference sample to each individual MS run. For example, in cancer studies, a mixed sample is typically
 generated by pooling tumor samples and/or normal adjacent tissues (NATs) from several related patients
 in equal protein amounts. The proteome of this mixed sample is then profiled in every MS run separately.
 reverse-zMAP module alleviates the influence of outliers by repeatedly making pairwise comparisons, which
 in the meanwhile allows the modeling of sample-specific mean-variance trend, but it requires a biologically
 identical reference sample in each MS run for a subsequent integration across MS runs.
 
# Workflow of zMAP

![Workflow of zMAP](https://github.com/guixiuqi/reverse-zMAP/blob/main/imgs/reverse_zmap_workflow.png "zMAP Workflow")

# Try it

A Web-based application of zMAP is provided at http://bioinfo.sibs.ac.cn/shaolab/zMAP. 


# Installation
## From source code
The scripts of reverse-zMAP  require no installation and can be used in-place. Just install the dependencies (see below)

```python
git clone https://github.com/guixiuqi/zMAP.git
cd ./reverse-zMAP
scriptPATH=./python_script
export scriptPATH
gmtPATH=./data/gmt
inputdataPATH=./data/test_data
```
## Python dependencies
```bat
conda install matplotlib  #version 3.6.2
conda install conda-forge::pandas #version 1.5.1
conda install anaconda::scipy #version 1.10.1
conda install anaconda::seaborn #version 0.12.2
conda install -c conda-forge scikit-learn #version 1.1.3
conda install conda-forge::lifelines #version 0.27.4
conda install conda-forge::statsmodels #version 0.13.5
conda install conda-forge::fastcluster #version 1.2.6

```

## R dependencies
```bat
library(GSVA)  #version 1.46.0
library(limma) #version 3.54.0
library(GSEABase,quietly=TRUE) #version 1.60.0
library(ConsensusClusterPlus) #version 1.62.0
library(MatrixEQTL) #version 2.3 
```


# Usage

## reverse-zMAP
reverse-zMAP models the mean-variance dependence associated with ILMS intensities and accordingly applies a variance-stabilizing z-transformation (z-statistic), 
which dramatically increases the comparability between samples generated by different MS runs. The z-statistic can be widely applied in downstream analyses.

Two input files provided by user:

1. Protein intensity file
   
    A tab-delimited file containing raw gene-level protein intensity with samples in columns, and gene symbols in rows.
Note:(1). The protein intensity matrix does not require normalization.(2). Sample names can only consist of letters, numbers, and underscores.

2. Sample information file

    Sample information file is a four-column, tab-delimited file with the first line identifying the columns. The column names are ```Sample_id```, ```MS_run```, ```Sample_condition```, and ```internal_ref```.

Use the command shown as below:

```bat
python $scriptPATH/reverse_zmap.py --intensity_data $inputdataPATH/small_data_protein_level_intensity.txt --sample_information $inputdataPATH/small_data_sample_info_add_internal_ref.txt --outdir $inputdataPATH/reverse_zMAP_results --window_size 400 --step_size 100 --percent 50 --method natural_cubic_spline
```
## Downstream analyses based on z-statistic

### Sample Quality Control

Performing principal component and hierarchical clustering analysis on the z-statistic matrix of samples provides a concise visualization of the overall impact of sample conditions and MS runs.
![QC of reverse-zMAP](https://github.com/guixiuqi/reverse-zMAP/blob/main/imgs/reverse-zMAPQC.png "reverseMAPQC")

Two input files provided by user:

1. Output file z_statistic_table.txt from reverse-zMAP

2. Sample information file

    Sample information file is a three-column, tab-delimited file with the first line identifying the columns. The column names are ```Sample_id```, ```Sample_condition```, and ```MS_run```.

Use the command shown as below:
```bat
python $scriptPATH/SampleQC.py --z_statistic_matrix $inputdataPATH/reverse_zMAP_results/z_statistic_table.txt --sample_info $inputdataPATH/small_data_sample_info_add_internal_ref.txt --outdir $inputdataPATH/SampleQC
```

### sample subgrouping

For delineating molecular subtypes at the protein level, this function is designed to select a subset of hypervariable proteins across samples. 
It then undertakes unsupervised clustering on the samples, offering quantitative evidence to ascertain both the number and composition of potential
 clusters within the dataset. For in-depth algorithmic insights, please consult the details provided in Consensus Clustering and R package ```ConsensusClusterPlus```.

Two input files provided by user:

1. Output file z_statistic_table.txt from reverse-zMAP

2. Sample information file

    Sample information file is a three-column, tab-delimited file with the first line identifying the columns. The column names are ```Sample_id```, ```Sample_condition```, and ```MS_run```.
   
Parameters:

1. Sample condition (```--sample_condition```)

   Perform clustering on specific condition of samples.

2. Top N(number) hypervariable proteins for subgrouping (```--top_n```)

   Proteins were ranked based on their variance across samples, and the z-statistic matrices of the top_n proteins were used for clustering. By default, the top_n is set to 3000.


Use the command shown as below:
```bat
python $scriptPATH/SampleSubgrouping.py --z_statistic_matrix $inputdataPATH/cervical_cancer_z_statistic.txt --sample_info $inputdataPATH/cervical_cancer_sample_info.txt --sample_condition Tumor_tissue --outdir $inputdataPATH/SampleSubgrouping --top_n 3000
```

### Gene set variation analysis
GSVA calculates gene set enrichment scores (GSVA scores) for each sample using the z-statistic matrix.
Differential expression analysis is then conducted on these GSVA scores using limma, aiming to identify differentially regulated pathways across sample groups. Finally, the differential pathway activities across sample groups are visualized using a heatmap.
![GSVA of zMAP](https://github.com/guixiuqi/reverse-zMAP/blob/main/imgs/zmap_gsva.png "zMAP_GSVA")

Two input files provided by user:

1. Output file z_statistic_table.txt from reverse-zMAP

2. Sample information file

    Sample information file is a three-column, tab-delimited file with the first line identifying the columns. The column names are ```Sample_id```, ```Sample_condition```, and ```MS_run```.

Parameters:

1. Top N(number) Pathways (```--top_n```)
 
   Extract a table of the top-ranked differentially regulated pathways across sample groups.

2. FDR (```--fdr```)

   BH-adjusted pvalue cutoff for significance.



Use the command shown as below:
```bat
python $scriptPATH/gsva.py --z_statistic_matrix $inputdataPATH/cervical_cancer_z_statistic.txt --sample_info $inputdataPATH/cervical_cancer_sample_info.txt --outdir $inputdataPAT
H/gsva --top_n 50 --fdr 0.05
```

### Association with clinical and molecular features
Samples were grouped using the z-statistic matrix derived from HVPs, with the aim of correlating proteomic subgroups
 to clinical and molecular features.We employed either the Chi-square test or Fisher's exact test to examine the association
 between proteomic subgroups and discrete sample features, selecting the appropriate test based on the number of categories
 within each feature. For continuous features, we utilized Student’s t-test or ANOVA.

ANOVA was further applied to identify differentially expressed proteins (DEPs) among sample groups.
Subsequently, hierarchical clustering of proteins, driven by the z-statistic matrix, revealed distinctive
expression signatures. Finally, pathway enrichment analysis was conducted for each set of proteins.


![AssClinicalMolecular](https://github.com/guixiuqi/reverse-zMAP/blob/main/imgs/AssClinicalMolecular.png "AssClinicalMolecular")

Two input files provided by user:

1. Output file z_statistic_table.txt from reverse-zMAP

2. Sample cluster file

    A comma-separated file with the first column containing sample names and the second column indicating the respective sample groups.

3. Clinical features(discrete variable name)

    This is a string where discrete features are comma-separated. For example, 'gender, stage'.

4. Clinical features(continuous variable name)

    This is a string where continuous features are comma-separated. For example, 'age, tumor_size'.

5. Color file


    Tab-delimited file with no column names. The first column represents discrete sample features, and the second column corresponds to the hexadecimal color code for each feature.

6. Colorbar file

    Tab-delimited file with no column names. The first column represents continuous sample features, and the second column corresponds to the colormaps in Matplotlib for each feature.

	
Parameters:

1. FDR  (```--fdr```)
 
   FDR cutoff used for identifying DEPs. The default FDR is 0.05.

2. Protein clusters (```--cluster_n```)

   Cluster number for DEPs.



Use the command shown as below:
```bat
python $scriptPATH/AssClinicalMolecular.py --z_statistic_matrix $inputdataPATH/cervical_cancer_z_statistic.txt --cluster_f $inputdataPATH/SampleSubgrouping/cluster_3.csv --clinical_info $inputdataPATH/cervical_cancer_clinical_info.txt --discrete 'Histology,Degree,Stage,Lymph_node,Abnormal_SCC' --continuous 'Tumor_size_cm,Age' --color_f $inputdataPATH/discrete_feature_color_RGB.txt --colorbar_f $inputdataPATH/continuous_feature_colorbar.txt --outdir $inputdataPATH/AssClinicalMolecular --fdr 0.05 --cluster_n 4
```

### Survival analysis
Plots the Kaplan-Meier survival curve and assesses the significance of survival differences among subgroups.


![Survival analysis](https://github.com/guixiuqi/reverse-zMAP/blob/main/imgs/AssClinicalMolecular.png "Survival analysis")

One input file provided by user:

1. Survival and group file

    This file is a four-column, tab-delimited file with the first line identifying the columns. The column names are Sample_id, survival_time,death_or_not, and group.

Use the command shown as below:
```bat
python $scriptPATH/sample_group_survival_analysis.py --input_file $inputdataPATH/survival_analysis_data.txt --outdir $inputdataPATH/SurvivalAnalysis
```

### Association with survival data

Cox proportional hazard regression model is used to identify prognostic markers based on z-statistic.
![GSVA of zMAP](https://github.com/guixiuqi/reverse-zMAP/blob/main/imgs/zmap_gsva.png "zMAP_GSVA")

Two input files provided by user:

1. Output file z_statistic_table.txt from reverse-zMAP
    

2. Survival file

	This file is a three-column, tab-delimited file with the first line identifying the columns. The column names are Sample_id, survival_time, and death_or_not.

Use the command shown as below:
```bat
python $scriptPATH/AssSurvival.py --survival_f $inputdataPATH/survival_analysis_data.txt --z_statistic_matrix $inputdataPATH/HCC_z_statistic_df.txt --outdir $inputdataPATH/AssSurvival
```

### Association with mutation data

To pick out the proteins that were associated with non-silent somatic mutations, QTL analysis is used. 
Somatic mutations detected in at least 10 more samples are included in the analysis (Users determine based on the number of samples). 
Then, we pick out the proteins whose expression was detected in at least half of the tumor samples. 
Finally, for each candidate gene-protein pair, we performed a regression of the protein expression 
against the mutation status of the gene using MatrixEQTL R package, by fitting the following linear 
model: Y = α + βX + γS + δA+ε. Here, Y was a vector of the z-statistics of the protein in tumor samples; 
X was the non-silent somatic mutation indicators of the gene; S and A referred to the sex and age variables, 
respectively, as well as other control variables; ε was a vector of independent and identically distributed 
noise variables; α, β, γ, and δ were unknown parameters. This model was fitted by applying the least squares method,
 and the null hypothesis β=0 was then tested by applying the t-test.
![GSVA of zMAP](https://github.com/guixiuqi/reverse-zMAP/blob/main/imgs/zmap_gsva.png "zMAP_GSVA")

Five input files provided by user:

1. Output file z_statistic_table.txt from reverse-zMAP
    

2. Mutation file

	This file is a three-column, tab-delimited file with the first line identifying the columns. The column names are Sample_id, survival_time, and death_or_not.
	
3. Covariates file

    The file includes additional covariates. Columns represent samples, while rows encompass various additional covariates, 
	including age, sex, and others. The values inside must be numerical.

4. Gene location file

	The file containing transcription start site information for all genes in the genome. 
	It contains three columns with the column names being gene, chromosome, and tss respectively.

5. Chromosome length file
    Tab-delimited file contains two columns, chromosome and length.

Parameters:

1. FDR  (```--fdr```)
 
   The default FDR is 0.05.


Use the command shown as below:
```bat
python $scriptPATH/AssMutation.py --mutation_f $inputdataPATH/snp_df_10.txt --z_statistic_matrix $inputdataPATH/HCC_z_statistic_df.txt --covariates_f $inputdataPATH/covariates_df.txt --gene_tss_location $inputdataPATH/gene_tss_location.txt --chr_length $inputdataPATH/hg19_chrom_sizes_remove_xy.txt --outdir $inputdataPATH/AssMutation --fdr 0.05
```




# Citation

To cite the `zMAP` tools in publications, please cite zMAP toolset: model-based analysis of large-scale proteomic data via a variance stabilizing z-transformation.


